reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1319 @@
+	OverlapPackets int
+}
+
+// ScatterGather is used to pass reassembled data and metadata of reassembled
+// packets to a Stream via ReassembledSG
+type ScatterGather interface {
+	// Returns the length of available bytes and saved bytes
+	Lengths() (int, int)
+	// Returns the bytes up to length
+	Fetch(length int) []byte
+	// Tell to keep from offset
+	KeepFrom(offset int)
+	// Return CaptureInfo of packet corresponding to given offset
+	CaptureInfo(offset int) gopacket.CaptureInfo
+	// Return some info about the reassembled chunks
+	Info() (direction TCPFlowDirection, start bool, end bool, skip int)
+	// Return some stats regarding the state of the stream
+	Stats() TCPAssemblyStats
+}
+
+// byteContainer is either a page or a livePacket
+type byteContainer interface {
+	Bytes() []byte
+	Length() int
+	ConvertToPages(*pageCache, int, AssemblerContext) (*page, *page, int)
+	CaptureInfo() gopacket.CaptureInfo
+	AssemblerContext() AssemblerContext
+	Release(*pageCache) int
+	Start() bool
+	End() bool
+	Seq() Sequence
+	IsPacket() bool
+}
+
+// Implements a ScatterGather
+type reassemblyObject struct {
+	all       []byteContainer
+	Skip      int
+	Direction TCPFlowDirection
+	saved     int
+	toKeep    int
+	// stats
+	queuedBytes    int
+	queuedPackets  int
+	overlapBytes   int
+	overlapPackets int
+}
+
+func (rl *reassemblyObject) Lengths() (int, int) {
+	l := 0
+	for _, r := range rl.all {
+		l += r.Length()
+	}
+	return l, rl.saved
+}
+
+func (rl *reassemblyObject) Fetch(l int) []byte {
+	if len(rl.all) == 1 && l == rl.all[0].Length() {
+		return rl.all[0].Bytes()
+	}
+	bytes := []byte{}
+	for _, bc := range rl.all {
+		bytes = append(bytes, bc.Bytes()...)
+	}
+	return bytes[:l]
+}
+
+func (rl *reassemblyObject) KeepFrom(offset int) {
+	rl.toKeep = offset
+}
+
+func (rl *reassemblyObject) CaptureInfo(offset int) gopacket.CaptureInfo {
+	current := 0
+	for _, r := range rl.all {
+		if current >= offset {
+			return r.CaptureInfo()
+		}
+		current += r.Length()
+	}
+	// Invalid offset
+	return gopacket.CaptureInfo{}
+}
+
+func (rl *reassemblyObject) Info() (TCPFlowDirection, bool, bool, int) {
+	return rl.Direction, rl.all[0].Start(), rl.all[len(rl.all)-1].End(), rl.Skip
+}
+
+func (rl *reassemblyObject) Stats() TCPAssemblyStats {
+	packets := int(0)
+	for _, r := range rl.all {
+		if r.IsPacket() {
+			packets++
+		}
+	}
+	return TCPAssemblyStats{
+		Chunks:         len(rl.all),
+		Packets:        packets,
+		QueuedBytes:    rl.queuedBytes,
+		QueuedPackets:  rl.queuedPackets,
+		OverlapBytes:   rl.overlapBytes,
+		OverlapPackets: rl.overlapPackets,
+	}
+}
+
+const pageBytes = 1900
+
+// TCPFlowDirection distinguish the two half-connections directions.
+//
+// TCPDirClientToServer is assigned to half-connection for the first received
+// packet, hence might be wrong if packets are not received in order.
+// It's up to the caller (e.g. in Accept()) to decide if the direction should
+// be interpretted differently.
+type TCPFlowDirection bool
```

Comment:
It's just an enum for the direction of the half-stream: we need to have exactly two different directions. We could change here the name to Dir1 and Dir2 for instance if the Client/Server names are too misleading. We can also add info (here ternary) in ScaterGather.Info() telling whether Dir1 is Client→Server or Server→Client or unknown.